### PR TITLE
Add cuDNN GAT and GATv2 Python APIs

### PR DESCRIPTION
**Pull request by tingyu66**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I reviewed the Hard Rules in the `AGENTS.md` for each directory this PR touches (see [root AGENTS.md § Reviewing a PR](https://github.com/NVIDIA/cudnn-frontend/blob/develop/AGENTS.md#reviewing-a-pr-human-or-agent)) and my changes comply, or I explain the exception below.
- [ ] I added GitHub labels: one `cat-*`, one or more `area:*` / `op:*`, and one `orig-*` (see [label list](https://github.com/NVIDIA/cudnn-frontend/labels)).

## Affected area

Python API or bindings

## Summary

- Add Python bindings and PyTorch autograd support for the cuDNN GNN GAT and GATv2 APIs.
- Expose the public `cudnn.gnn.mha_gat` and `cudnn.gnn.mha_gat_v2` operations.
- Support homogeneous and bipartite graphs through either a single feature tensor or separate source and destination feature tensors.
- Support edge features, caller-provided dropout masks, optional attention-weight outputs, configurable activations, and concatenated or averaged attention heads.
- Add deterministic backward execution using reverse-CSC graph metadata.
- Add gradient-precision controls:
  - GAT: `high_precision_dgrad` and `high_precision_wgrad`
  - GATv2: `high_precision_grad`
- Extend `CscGraph` with optional reverse-CSC metadata and a `with_reverse_csc()` helper.
- Add API documentation, public API-index entries, shared GNN test fixtures, and L0–L2 coverage for GAT, GATv2, and `agg_simple`.

## Why

The cuDNN backend now provides native GAT and GATv2 operations in 9.28, but they were not accessible through the cuDNN Frontend Python package.

This change exposes those operations through Python APIs that integrate with PyTorch autograd and follow the existing `cudnn.gnn` conventions. It also represents bipartite graphs explicitly, exposes the backend’s deterministic and gradient-precision controls, and keeps reverse-CSC metadata on the reusable graph descriptor.

## Related issues

None.

## API and compatibility impact

This PR adds the following public Python APIs:

- `cudnn.gnn.mha_gat`
- `cudnn.gnn.mha_gat_v2`
- `cudnn.gnn.CscGraph.has_reverse_csc`
- `cudnn.gnn.CscGraph.with_reverse_csc`

`CscGraph` gains the optional, append-only fields `csc_rev_offsets` and `map_rev_to_coo`. Existing construction patterns remain supported.

The operations support CUDA FP32, FP16, and BF16 features and weights, with INT32 or INT64 CSC indices. They require:

- cuDNN 9.28 or newer headers when building the bindings
- a backend containing the corresponding GNN entry points
- SM 8.0 or newer
- a non-Windows platform

Deterministic backward requires reverse-CSC metadata. `CscGraph.with_reverse_csc()` can construct it when it is not supplied by the application.

For FP16 and BF16 inputs, the high-precision options request FP32 gradient computation. They have no effect for FP32 inputs. GAT does not support enabling `high_precision_dgrad` without also enabling `high_precision_wgrad`, matching the backend’s supported gradient-type combinations.

No existing public API is removed.

## Testing

Tested against the updated cuDNN backend debug build (`linux-x86_64-13.2--debug`).

```bash
cd test/python
pytest gnn -m L0

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added GAT and GATv2 multi-head attention operations for homogeneous and bipartite CSC graphs.
  * Added support for configurable activations, multiple heads, edge features, dropout, attention-weight returns, deterministic backward execution, and gradient precision controls.
  * Added reverse-graph metadata support for deterministic operations.
  * Added Python API exports and Frontend API documentation.

* **Documentation**
  * Expanded aggregation documentation with usage requirements and behavior details.

* **Tests**
  * Added comprehensive coverage for attention operations, graph variants, gradients, validation, compilation, dropout, and deterministic execution.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by tingyu66** [comment]
@coderabbitai help

**Comment by tingyu66** [comment]
@coderabbitai full review

File: python/cudnn/gnn/graph.py [hunk 1]
```diff
@@ -15,18 +16,49 @@ class CscGraph:
...
+        constructed on the graph device using the current PyTorch stream.
+        """
+        if self.has_reverse_csc:
+            return self
+
+        from ._utils import validate_csc_graph
+
+        _, num_edges = validate_csc_graph(self.offsets, self.indices, self.map_csc_to_coo, self.num_src_nodes)
+        counts = torch.zeros((self.num_src_nodes,), device=self.indices.device, dtype=self.indices.dtype)
+        counts.scatter_add_(0, self.indices.long(), torch.ones_like(self.indices))
```
[the remark this answers is not in the document]
Comment (tingyu66) [reply python/cudnn/gnn/graph.py:56]: check again

File: test/python/api_index/api_index.txt [hunk 2]
```diff
@@ -326,11 +326,27 @@ cudnn.get_last_error_string
 cudnn.get_stream
 cudnn.gnn
 cudnn.gnn.CscGraph
+cudnn.gnn.CscGraph.csc_rev_offsets
+cudnn.gnn.CscGraph.has_reverse_csc
 cudnn.gnn.CscGraph.map_csc_to_coo
+cudnn.gnn.CscGraph.map_rev_to_coo
 cudnn.gnn.CscGraph.num_dst_nodes
+cudnn.gnn.CscGraph.with_reverse_csc
```
[the remark this answers is not in the document]
Comment (tingyu66) [reply test/python/api_index/api_index.txt]: check again

File: python/cudnn/gnn/graph.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tingyu66) [reply python/cudnn/gnn/graph.py:56]: indices is a CUDA tensor. Verifying every value is in [0, num_src_nodes) would require:
1. Launching one or more CUDA kernels for comparisons/reduction.
2. Reading the reduction result on the CPU to decide whether to raise.
3. Synchronizing the CUDA stream.
That makes every validate_csc_graph() call more expensive and violates the frontend’s Rule 3: validation must not read device data back to the host.

File: test/python/api_index/api_index.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (tingyu66) [reply test/python/api_index/api_index.txt]: fixed in [1d14198](https://github.com/NVIDIA/cudnn-frontend/pull/1021/commits/1d14198a7dec6b1e3afdef1643ca862b04091cdf)

**Comment by tingyu66** [comment]
@coderabbitai full review

File: test/python/api_index/api_index.txt [hunk 3]
```diff
@@ -326,11 +326,27 @@ cudnn.get_last_error_string
...
+cudnn.gnn.CscGraph.has_reverse_csc
 cudnn.gnn.CscGraph.map_csc_to_coo
+cudnn.gnn.CscGraph.map_rev_to_coo
 cudnn.gnn.CscGraph.num_dst_nodes
 cudnn.gnn.CscGraph.num_edges
 cudnn.gnn.CscGraph.num_indices
+cudnn.gnn.CscGraph.with_reverse_csc
 cudnn.gnn.agg_simple
+cudnn.gnn.gat
+cudnn.gnn.gat_v2
```
[the remark this answers is not in the document]
Comment (tingyu66) [reply test/python/api_index/api_index.txt]: fixed in [c79c553](https://github.com/NVIDIA/cudnn-frontend/pull/1021/commits/c79c55361c3d4993786f00b41274781368b492c0)

File: python/gnn.cpp [hunk 4]
```diff
@@ -191,6 +200,320 @@ init_gnn_submodule([[maybe_unused]] py::module_ &m) {
...
+                                                        reinterpret_cast<const void *>(attn_weights),
+                                                        reinterpret_cast<const float *>(dropout_mask),
+                                                        reinterpret_cast<void *>(output),
+                                                        reinterpret_cast<void *>(sm_scores),
+                                                        node_feat_dim,
+                                                        edge_feat_dim,
+                                                        &params,
+                                                        static_cast<cudnnDataType_t>(data_type));
+              throw_if_gnn_failed(status, "cudnnGnnMhaGatForward");
+          });
```
[the remark this answers is not in the document]
Comment (tingyu66) [reply python/gnn.cpp]: fixed in [c79c553](https://github.com/NVIDIA/cudnn-frontend/pull/1021/commits/c79c55361c3d4993786f00b41274781368b492c0)

**Comment by tingyu66** [comment]
@cudnn-ci-bot run oss

**Review by vedaanta — COMMENTED** [review]
**Codex bot review · model: `gpt-5.6-sol`**

Reviewed `c79c55361c3d` against base `9b9f32781d0c`.

Thank you for your contribution! I reviewed the full GAT/GATv2 Python API, binding, autograd, graph-metadata, documentation, and test changes. I found two approval-blocking P1 issues: accepted noncontiguous feature layouts can silently produce incorrectly interpreted gradients, and compile-time cuDNN 9.28 guards permanently remove the new bindings from artifacts built with older supported headers. Please address both before merging. The exact-head OSS pipeline also has two new Linux release-build failures; its Python test jobs remained manual.

**[P1 · blocks approval] Allocate backend gradient outputs with contiguous layouts** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/c79c55361c3d4993786f00b41274781368b492c0/python/cudnn/gnn/mha_gat.py#L447))

The public validation accepts noncontiguous feature tensors and forward explicitly normalizes them with `.contiguous()`, but backward allocates `grad_src`, `grad_dst`, and `grad_edge` with `torch.empty_like`. For a dense transpose, `empty_like` preserves the transposed strides. The `cudnnGnnMhaGatBackward` ABI receives only raw pointers and dimensions, with no stride metadata, so its contiguous writes are then interpreted through a different logical layout. The same pattern appears in GATv2 at `mha_gat_v2.py:420-423` and in both fake backward allocators. Please either allocate contiguous gradient buffers (and make the fake implementation report the same layouts) or reject noncontiguous inputs. This P1 blocks approval because a normal, accepted tensor view can silently return wrong gradients rather than raising an error.

Evidence: A focused CUDA allocation probe passed a `(2, 4)` transposed feature with stride `(1, 2)` through the current GAT backward allocation path and received `grad_src` with the same noncontiguous stride. Writing the backend's expected contiguous sequence 0..7 into that storage makes the returned logical tensor `[[0,2,4,6],[1,3,5,7]]`, rather than `[[0,1,2,3],[4,5,6,7]]`. Source inspection confirms neither C++ backward binding accepts strides. Attribution: Introduced by this PR's GAT and GATv2 backward implementations.

**[P1 · blocks approval] Keep the 9.28 bindings in artifacts built with older headers** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/c79c55361c3d4993786f00b41274781368b492c0/include/cudnn_frontend_shim.h#L717))

The `#if CUDNN_VERSION >= 92800` block removes all four shim declarations, and the matching guard in `python/gnn.cpp:204` removes their pybind exports. This contradicts the include-layer hard rule to declare newer-cuDNN APIs unconditionally and gate them using `detail::get_backend_version()` at runtime. A wheel built against an older supported cuDNN 9.x header therefore permanently lacks these bindings; installing it next to a 9.28 backend cannot enable the advertised public APIs. Please keep a version-neutral binding ABI in the artifact and return a clear not-supported result when the loaded backend is older than 9.28. This P1 blocks approval because it makes the entire new API unavailable on a normal supported build/install path, and rebuilding the wheel is the only workaround.

Evidence: A clean Release extension build of this head against cuDNN 9.20 headers completed, but import inspection showed `gnn_mha_gat_forward`, `gnn_mha_gat_backward`, `gnn_mha_gat_v2_forward`, and `gnn_mha_gat_v2_backward` were all absent while `cudnn.gnn.mha_gat` remained publicly importable and failed at `_require_backend`. Because the pybind method table is fixed at build time, swapping in a newer runtime library cannot add those exports. Attribution: Introduced by this PR's new 9.28 shim and pybind guards.

Validation: Inspected the complete 17-file diff and traced the public API through validation, custom-op schemas, fake implementations, autograd setup/backward, pybind calls, shim calls, graph metadata, docs, and tests. Built `_compiled_module` in Release mode from this head against cuDNN 9.20 headers and imported the resulting package to inspect its exported symbol set and public GAT failure mode. Ran a focused CUDA layout/allocation probe on an A100 with torch 2.13.0+cu130, including a known-value contiguous-storage write demonstrating the transposed logical gradient. Ran `py_compile` on the changed GNN Python implementation files. Exact-head OSS pipeline 69100936 reported 17 passed, 2 new Linux release-build failures, and 6 manual jobs; the API-index, dev Linux builds, Windows release build, analysis, sanitizers, and SM80/SM90/SM100 C++ tests passed. GitHub style and documentation checks passed on the reviewed head.

Limitations: A cuDNN 9.28 backend was not available in the local environment, so I did not execute the new GAT kernels numerically; the exact-head Python test jobs were manual and did not run. The internal logs for the two exact-head Linux release-build failures were not accessible, so I did not attribute those failures to a specific source defect.

I am withholding approval for the high-risk P1 findings above and leaving a COMMENT. Please address them before merging.

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `a23ab5a7bb25` against base `dae3a33463d8`.

Thank you for your contribution! The noncontiguous-gradient fix looks correct: both real and fake backward allocation paths now produce compact gradients, and the updated bipartite tests compare all feature gradients against the reference. My focused old/new allocation probe fails on the old revision and passes on this head. The earlier build-time binding availability issue remains unchanged. I also traced the two previously reported release-build failures to the new reverse-CSC member guard; please address both build compatibility issues before merging.

**[P1 · blocks approval] Match the reverse-CSC field guard to the headers that provide it** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/a23ab5a7bb2588038add8b3472b4f9cea37cdab5/python/gnn.cpp#L36))

The new `CUDNN_VERSION >= 92700` branch unconditionally assigns `graph.cscRevOffsets`, but the headers used by both Linux release-build jobs do not provide that member. Both jobs fail at line 37 with `cudnnGnnCscGraph_t has no member named cscRevOffsets`. This helper is compiled independently of the 9.28 GAT binding guard, so it breaks the whole Python extension build even when GAT itself would be excluded. Please use the actual header/API availability boundary for the reverse-CSC fields and verify both the older supported build and the new GAT build. This is an approval-blocking P1 because it fails the normal wheel build on both Linux architectures, rather than just reporting that an optional operation is unsupported.

Evidence: Inspected pipeline 69100936 jobs 450025805 and 450025809: both fail compiling python/gnn.cpp:37 with the same missing-member diagnostic. That pipeline tested c79c553; python/gnn.cpp is byte-identical at a23ab5a. The assignment and its 92700 guard are new relative to the PR merge base. Attribution: Introduced by this PR's make_csc_graph extension; present before the latest Python layout fix.

- **[P1 · blocks approval] Still open:** [Keep the 9.28 bindings in artifacts built with older headers](https://github.com/NVIDIA/cudnn-frontend/pull/1021#pullrequestreview-5271396140).

Validation: Reviewed all six changed files since c79c553, including the real/fake allocation changes, updated transposed bipartite gradient assertions, documentation and navigation slug. Executed the exact real/fake allocation function definitions in a CPU harness with CUDA validation/backend calls stubbed. Covered GAT and GATv2, transposed source/destination/edge tensors, strided weights, optional edges, FP16/FP32 gradients and both deterministic settings. The old revision failed the compact-layout/packed-write checks in 32/32 cases; this head passed 32/32. Independently read both prior-head Linux release-build logs and checked that the failing native source is unchanged at the reviewed head. Confirmed the reverse-CSC assignment is new relative to the merge base.

Limitations: The allocation probe validates buffer layout and simulated packed writes, not native GAT numerical behavior. A cuDNN 9.28 backend was unavailable locally, so I did not rerun the new GPU regressions or measure GAT performance. Pipeline 69100936 tested c79c553, not this head. Its unchanged native-source failure remains relevant, but it is not a current-head CI result. No additional broad CI was triggered while these build issues remain open.

I am withholding approval for the high-risk P1 findings above and leaving a COMMENT. Please address them before merging.

**Review by YangXu1990uiuc — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `2bab7470dc71` against base `a5d61f45c08d`.

Thank you for your contribution! Both build-compatibility blockers from my previous review are resolved. The frontend-owned ABI types compile against older headers, and GAT/GATv2 remain exported with runtime version checks. I verified the important upgrade path directly: the extension built with cuDNN 9.26 headers runs the GAT/GATv2 tests successfully with a real 9.28 backend. The earlier noncontiguous-gradient fix is retained. No remaining approval-blocking findings.

Validation: Reviewed all five files changed since a23ab5a7, including ABI layout checks, all four dynamic/linked GAT call paths, version errors, documentation and test guards. Checked intervening base changes for the affected contracts. Built the complete exact-head Python extension against cuDNN 9.26.0.51 headers. The changed binding also compiles with actual 9.23.1 and 9.28 headers under the project's warnings-as-errors flags, including its ABI assertions. H100, Torch 2.13, cuDNN 9.26: binding-export and aggregation forward/backward tests passed, 80 passed and 2 L2 cases deselected. All four actual GAT/GATv2 native entry points reject 9.26 with an explicit requires-9.28/loaded-9.26 error before CUDA initialization. Using the same 9.26-built extension with a real cuDNN 9.28 backend on H100: GAT and GATv2 suites with explicit L0 or L1 selection finished with 188 passed, 4 L2 cases deselected. This includes noncontiguous bipartite gradients, deterministic paths, optional edges, activations and precision controls. Exact-head Style and Docs workflows passed.

Limitations: Independent numerical validation was on H100; other architectures and the current-head release-wheel CI remain separate checks. No performance comparison was run for this binding-compatibility follow-up.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.

**Comment by YangXu1990uiuc** [comment]
The build-compatibility follow-up is reviewed and the focused local checks passed. I am requesting `python_tests` for `2bab7470dc71`: that target collects the GNN suite and also rechecks the release builds that failed previously. The earlier `oss` run did not execute these GNN tests. This CI run is additional, pending validation.

Codex bot review — model `gpt-6-astra`.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run python_tests

**Comment by YangXu1990uiuc** [comment]
**Codex bot review · model: `gpt-6-astra`**

CI follow-up for `2bab7470dc71`: pipeline **69610589** completed successfully. All five platform builds and all six selected dev/release Python lanes (SM80, SM90 and SM100) passed, including both release Linux builds that failed on the previous head. The inspected development lanes used cuDNN 9.28; the release lane used 9.27.

This confirms the CI build-compatibility fixes. The independent 188-case GAT/GATv2 numerical run on cuDNN 9.28 is recorded separately in the review above.

Codex bot review — model `gpt-6-astra`.
